packages/db/src/core/tokens.ts

```diff
@@ -95,24 +102,35 @@ class ManagedRemoteAppToken implements ManagedAppToken {
 		);
 	}
 
+	tokenIsValid() {
+		return new Date() > this.expires;
+	}
+
+	createRenewTimer() {
+		return setTimeout(() => this.renew(), (1000 * 60 * this.ttl) / 2);
+	}
+
 	async renew() {
 		clearTimeout(this.renewTimer);
 		delete this.renewTimer;
-		try {
+
+		if(this.tokenIsValid()) {
 			const response = await this.fetch('/auth/cli/token-renew', {
 				token: this.token,
 				projectId: this.projectId,
 			});
 			if (response.status === 200) {
-				this.renewTimer = setTimeout(() => this.renew(), (1000 * 60 * this.ttl) / 2);
+				this.expires = getExpiresFromTtl(this.ttl);
+				this.renewTimer = this.createRenewTimer();
 			} else {
 				throw new Error(`Unexpected response: ${response.status} ${response.statusText}`);
-			}
-		} catch (error: any) {
-			const retryIn = (60 * this.ttl) / 10;
-			// eslint-disable-next-line no-console
-			console.error(`Failed to renew token. Retrying in ${retryIn} seconds.`, error?.message);
-			this.renewTimer = setTimeout(() => this.renew(), retryIn * 1000);
+			}	
+		} else {
+			const { token, ttl } = await ManagedRemoteAppToken.createToken(this.session, this.projectId);
```

Comment:
if token renewal has failed I think you can assume getting a new token would fail as well and just exit instead, with a helpful error message like "token has expired and attempts to renew it have failed, please try again"

That would prevent the class of errors where this handling logic fails and swallows the original error (I think in this case, the user would get a "could not create a new token" error if this fails when really the error was around renewal and not creation).

Reply:
done, thanks